test/iterator.js

```diff
@@ -0,0 +1,359 @@
+    },
+  };
+
+  iterator.forEach(obj.fn, obj);
+
+  test.strictSame(obj.sum, 10);
+  test.end();
+});
+
+api.metatests.test('Iterator.reduce', test => {
+  const iterator = new Iterator(array);
+  const sum = iterator.reduce((acc, current) => acc + current, 0);
+
+  test.strictSame(sum, 10);
+  test.end();
+});
+
+api.metatests.test('Iterator.reduce with no initialValue', test => {
+  const iterator = new Iterator(array);
+  const sum = iterator.reduce((acc, current) => acc + current);
+
+  test.strictSame(sum, 10);
+  test.end();
+});
+
+api.metatests.test(
+  'Iterator.reduce with no initialValue on consumed iterator',
+  test => {
+    const iterator = new Iterator(array);
+    test.throws(() => {
+      iterator.reduce(() => {});
+      iterator.reduce((acc, current) => acc + current);
+    }, new TypeError('reduce of consumed iterator with no initial value'));
+
+    test.end();
+  }
+);
+
+api.metatests.test('Iterator.map', test => {
+  const iterator = new Iterator(array);
+  const sum = iterator
+    .map(value => value * 2)
+    .reduce((previous, current) => previous + current, 0);
+
+  test.strictSame(sum, 20);
+  test.end();
+});
+
+api.metatests.test('Iterator.map with thisArg', test => {
+  const iterator = new Iterator(array);
+  const obj = {
+    multiplier: 2,
+    mapper(value) {
+      return value * this.multiplier;
+    },
+  };
+
+  const sum = iterator
+    .map(obj.mapper, obj)
+    .reduce((previous, current) => previous + current, 0);
+
+  test.strictSame(sum, 20);
+  test.end();
+});
+
+api.metatests.test('Iterator.filter', test => {
+  const iterator = new Iterator(array);
+  const sum = iterator
+    .filter(value => !(value % 2))
+    .reduce((previous, current) => previous + current, 0);
+
+  test.strictSame(sum, 6);
+  test.end();
+});
+
+api.metatests.test('Iterator.filter with thisArg', test => {
+  const iterator = new Iterator(array);
+  const obj = {
+    divider: 2,
+    predicate(value) {
+      return !(value % this.divider);
+    },
+  };
+
+  const sum = iterator
+    .filter(obj.predicate, obj)
+    .reduce((previous, current) => previous + current, 0);
+
+  test.strictSame(sum, 6);
+  test.end();
+});
+
+api.metatests.test('Iterator.flat', test => {
+  const array = [[[[1], 2], 3], 4];
+  const flatArray = [1, 2, 3, 4];
+  test.strictSame(
+    iter(array)
+      .flat(3)
+      .toArray(),
+    flatArray
+  );
+  test.end();
+});
+
+api.metatests.test('Iterator.flat with no depth', test => {
+  const array = [[[[1], 2], 3], 4];
+  const flatArray = [[[1], 2], 3, 4];
+  test.strictSame(
+    iter(array)
+      .flat()
+      .toArray(),
+    flatArray
+  );
+  test.end();
+});
+
+api.metatests.test('Iterator.flatMap', test => {
+  const array = [1, 2, 3];
+  const result = [1, 1, 2, 2, 3, 3];
+  test.strictSame(
+    iter(array)
+      .flatMap(element => [element, element])
+      .toArray(),
+    result
+  );
+  test.end();
+});
+
+api.metatests.test('Iterator.flatMap', test => {
```

Comment:
It should probably be called `'Iterator.flatMap with thisArg'`.

Reply:
Fixed.